File: Array/018. 4Sum.cpp
/*
18. 4Sum

Given an array nums of n integers and an integer target, are there elements a, b, c, 
and d in nums such that a + b + c + d = target? Find all unique quadruplets in the array which gives the sum of target.

Note:

The solution set must not contain duplicate quadruplets.

Example:

Given array nums = [1, 0, -1, 0, -2, 2], and target = 0.

A solution set is:
[
  [-1,  0, 0, 1],
  [-2, -1, 1, 2],
  [-2,  0, 0, 2]
]


*/

//先比后跳
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if(nums.size()<4) return {};
        vector<vector<int>>res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i<nums.size()-3; i++){
            if(nums[i]+nums[i+1]+nums[i+2]+nums[i+3]>target) break;
            for(int j = i+1; j<nums.size()-2; j++){
                for(int k = j+1, m = nums.size()-1; k<m; ){
                    int temp = nums[i] + nums[j] + nums[k] + nums[m];
                    if(temp == target){
                        res.push_back({nums[i],nums[j],nums[k],nums[m]});
                        while(k<m && nums[k+1] == nums[k]) k++;
                        while(k<m && nums[m-1] == nums[m]) m--;
                        k++; m--;
                    }else if(temp < target){
                        while(k<m && nums[k+1] == nums[k]) k++;
                        k++;
                    }else{
                        while(k<m && nums[m-1] == nums[m]) m--;
                        m--;
                    }
                }
                while(j<nums.size()-3 && nums[j] == nums[j+1]) j++;
            }
            while(i<nums.size()-4 && nums[i+1] == nums[i]) i++;
        }
        return res;
    }
};


//先跳后比（指的是逻辑顺序，而不是实际顺序）
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if(nums.size()<4) return {};
        vector<vector<int>>res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i<nums.size()-3; i++){
            if(nums[i]+nums[i+1]+nums[i+2]+nums[i+3]>target) break;
            if(i>0 && nums[i] == nums[i-1]) continue;
            for(int j = i+1; j<nums.size()-2; j++){
                if(j>i+1 && nums[j] == nums[j-1]) continue;
                for(int k = j+1, m = nums.size()-1; k<m; ){
                    int temp = nums[i] + nums[j] + nums[k] + nums[m];
                    if(k > j+1 && nums[k] == nums[k-1])  k++;
                    else if(m < nums.size()-1 && nums[m] == nums[m+1]) m--;
                    else if(temp == target){
                        res.push_back({nums[i],nums[j],nums[k],nums[m]});
                        k++; m--;
                    }else if(temp < target){
                        k++;
                    }else{
                        m--;
                    }
                }
            }
        }
        return res;
    }
};


// Time:  O(n^4)
// Space: O(n^2)
class Solution2 {
public:
    vector<vector<int> > fourSum(vector<int> &num, int target) {
        vector<vector<int>> ans;
        if (num.size() < 4)  {
            return ans;
        }
        sort(num.begin(), num.end());
        unordered_multimap<int, pair<int, int>> cache;

        for (int i = 0; i < num.size(); ++i) {
            for (int j = i + 1; j < num.size(); ++j) {
                cache.emplace(num[i] + num[j], make_pair(i, j));
            }
        }

        for (auto i = cache.begin(); i != cache.end(); ++i) {
            auto a = i->second.first;
            auto b = i->second.second;
            auto range = cache.equal_range(target - i->first);
            for (auto j = range.first; j != range.second; ++j) {
                auto c = j->second.first;
                auto d = j->second.second;
                if (b < c) {
                    ans.push_back({num[a], num[b], num[c], num[d]});
                }
            }
        }
        sort(ans.begin(), ans.end());
        ans.erase(unique(ans.begin(), ans.end()), ans.end());
        return ans;
    }
};

File: Array/018. 4Sum.py
"""
18. 4Sum


Given array nums = [1, 0, -1, 0, -2, 2], and target = 0.

A solution set is:
[
  [-1,  0, 0, 1],
  [-2, -1, 1, 2],
  [-2,  0, 0, 2]
]

"""

class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        for ind, v1 in enumerate(nums[:-3]):
            if v1*4 > target: break
            if ind > 0 and nums[ind-1] == v1: continue
            for ind2, v2 in enumerate(nums[ind+1:-2],ind+1):
                if v1 + v2*3 > target: break
                if ind2 > ind + 1 and nums[ind2-1] == v2: continue
                i, j = ind2 + 1, len(nums)-1
                while i < j: 
                    tmp = v1 + v2 + nums[i] + nums[j]
                    if tmp == target:
                        res.append([v1, v2, nums[i], nums[j]])
                        while i + 1 < j and nums[i+1] == nums[i]:i+=1
                        while i < j - 1 and nums[j-1] == nums[j]: j-=1 
                        i+=1; j-=1
                    elif tmp <target: 
                        i+=1
                    else: 
                        j-=1
        return res 

#backtrack
class Solution:
    def fourSum(self, nums, target):
        def findNsum(nums, target, N, result, results):
            if len(nums) < N or N < 2 or target < nums[0]*N or target > nums[-1]*N:  # early termination
                return
            if N == 2: # two pointers solve sorted 2-sum problem
                l,r = 0,len(nums)-1
                while l < r:
                    s = nums[l] + nums[r]
                    if s == target:
                        results.append(result + [nums[l], nums[r]])
                        l += 1
                        while l < r and nums[l] == nums[l-1]:
                            l += 1
                    elif s < target:
                        l += 1
                    else:
                        r -= 1
            else: # recursively reduce N
                for i in range(len(nums)-N+1):
                    if i == 0 or (i > 0 and nums[i-1] != nums[i]):
                        findNsum(nums[i+1:], target-nums[i], N-1, result+[nums[i]], results)

        results = []
        findNsum(sorted(nums), target, 4, [], results)
        return result

#把4 sum 变成两个two sum
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        hash = {}
        res = set()
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] not in hash:
                    hash[nums[i] + nums[j]] = [(i,j)]
                else:
                    hash[nums[i] + nums[j]].append((i,j))
        for i in range(len(nums)):
            for j in range(i+1, len(nums)-2):
                if target - nums[i] - nums[j] in hash:
                    for k in hash[target - nums[i] - nums[j]]:
                        if k[0] > j:
                            res.add((nums[i], nums[j], nums[k[0]], nums[k[1]]))
        return [list(i) for i in res]